src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsBuilderExtensions.cs

```diff
@@ -58,22 +61,72 @@ public static IWebJobsBuilder UseHostId(this IWebJobsBuilder builder, string hos
 
         public static IWebJobsBuilder UseWebJobsStartup<T>(this IWebJobsBuilder builder) where T : IWebJobsStartup, new()
         {
-            return builder.UseWebJobsStartup(typeof(T));
+            return builder.UseWebJobsStartup<T>(NullLogger.Instance);
+        }
+
+        public static IWebJobsBuilder UseWebJobsStartup<T>(this IWebJobsBuilder builder, ILogger logger) where T : IWebJobsStartup, new()
+        {
+            return builder.UseWebJobsStartup(typeof(T), logger);
         }
 
         public static IWebJobsBuilder UseWebJobsStartup(this IWebJobsBuilder builder, Type startupType)
+        {
+            return builder.UseWebJobsStartup(startupType, NullLogger.Instance);
+        }
+
+        public static IWebJobsBuilder UseWebJobsStartup(this IWebJobsBuilder builder, Type startupType, ILogger logger)
         {
             if (!typeof(IWebJobsStartup).IsAssignableFrom(startupType))
             {
                 throw new ArgumentException($"The {nameof(startupType)} argument must be an implementation of {typeof(IWebJobsStartup).FullName}");
             }
 
             IWebJobsStartup startup = (IWebJobsStartup)Activator.CreateInstance(startupType);
-            startup.Configure(builder);
 
+            if (logger == NullLogger.Instance) 
+            {
+                startup.Configure(builder);
+            }
+            else
+            {
+                ConfigureAndLogUserConfiguredServices(startup, builder, logger);      
+            }
             return builder;
         }
 
+        private static void ConfigureAndLogUserConfiguredServices(IWebJobsStartup startup, IWebJobsBuilder builder, ILogger logger)
+        {
+            ITrackingServiceCollection tracker = builder.Services as ITrackingServiceCollection;
+
+            if (tracker != null)
+            {
+                startup.Configure(builder);
+
+                foreach (ServiceDescriptor service in tracker.Diff)
+                {
+                    string serviceName = service.ServiceType.FullName;
+                    string implementation = "";
+                    if (service.ImplementationInstance != null)
+                    {
+                        implementation = service.ImplementationInstance.GetType().FullName + "[instance]";
+                    }
+                    else if (service.ImplementationType != null)
+                    {
+                        implementation = service.ImplementationType.FullName;
+                    }
+                    else if (service.ImplementationFactory != null)
+                    {
+                        implementation = "Using ImplementationFactoy";
+                    }
+                 
+                    string lifetime = service.Lifetime.ToString(); 
+                    logger.LogInformation("Service type: {serviceName}, Implementation: {implementation}, Registered with lifetime: {lifetime}", serviceName, implementation, lifetime);
```

Comment:
Couple of suggestions on the message:
1. Instead of `Implementation: {implementation}`, you can do something like `{registeredServiceString}` and when you build it up, make the whole string be `Instance: {instanceType}` or `Implementation: {implemenationType}` or `Factory`.
2. My opinion -- I think we should compact this and format all of them into a single message using a `StringBuilder`. It'll be easier to find in the logs.

Something like: 
```
Services registered by external startup:
  {serviceName}, {registeredServiceString}, {lifetime}
  {serviceName}, {registeredServiceString}, {lifetime}
  {serviceName}, {registeredServiceString}, {lifetime}
```

Reply:
I like the suggestion. The one thing I'd add is the external startup type name to the header.